scripts/Deploy-CustomScriptsExtension.ps1

```diff
@@ -0,0 +1,179 @@
+        [Parameter(Mandatory=$true)][string]$ResourceGroupName,
+        [Parameter(Mandatory=$true)][string]$AadWebClientId,
+        [Parameter(Mandatory=$true)][string]$AadWebClientAppKey,
+        [Parameter(Mandatory=$true)][string]$AadTenantId,
+        [Parameter(Mandatory=$true)][string]$TemplateFile,
+        [Parameter(Mandatory=$true)][string]$TemplateParameterFile,
+        [Parameter(Mandatory=$false)][switch]$Upgrade
+     )
+
+##  Function: LogMessage
+##
+##  Purpose: Write a message to a log file
+##
+##  Input: 
+##      Message          - string - message to write
+##      LogType          - string - message type
+##      Foregroundcolor  - string - color of the output for Log-Messageonly
+##
+##  Ouput: null
+function Log-Message
+{
+    param(
+            [Parameter(Mandatory=$false)][object]$Message,
+            [Parameter(Mandatory=$false)][ValidateSet("Verbose","Output", "Host", "Error", "Warning")][string]$LogType="Host",
+            [Parameter(Mandatory=$false)][string]$Foregroundcolor = "White",
+            [Parameter(Mandatory=$false)][string]$Context = "",
+            [Parameter(Mandatory=$false)][switch]$NoNewLine,
+            [Parameter(Mandatory=$false)][switch]$ClearLine,
+            [Parameter(Mandatory=$false)][switch]$SkipTimestamp
+         )
+
+    
+    # append header to identify where the call came from for debugging purposes
+    if ($Context -ne "")
+    {
+        $Message = "$Context - $Message";
+    }
+
+    # if necessary, prepend a blank line
+    if ($ClearLine -eq $true)
+    {
+        $logTime = [System.Environment]::NewLine
+    }
+
+    # prepend log time
+    $logTime += "[$(get-date -format u)]";
+
+    if($NoNewLine -eq $false -and $SkipTimestamp -eq $false)
+    {
+        $logLine = "$logTime :: $Message";
+    }
+    else
+    {
+        $logLine = $Message;
+    }
+
+    switch($LogType)
+    {
+        "Verbose" {  Write-Verbose $logLine; }
+        "Output"  {  Write-Output $logLine ; }
+        "Host"    {  Write-Host $logLine -ForegroundColor $ForegroundColor -NoNewline:$NoNewLine; }
+        "Error"   {  Write-Error $logLine; }
+        "Warning" {  Write-Warning $logLine ; }
+        default   {  Write-Host $logLine -ForegroundColor $ForegroundColor -NoNewline:$NoNewLine; }
+    }
+}
+
+$jobs = @()
+
+# if upgrade is set, we skip deleting the existing extensions
+if ($Upgrade -eq $false)
+{
+
+    # get all VMs
+    Log-Message "Getting all VMs in the '$ResourceGroupName' resource group";
+    $vms = Get-AzureRmVM -ResourceGroupName $ResourceGroupName -Verbose
+
+
+    # iterate each VM and remove any existing custom scription extension
+    foreach($vm in $vms)
+    {
+        foreach($extension in $vm.Extensions)
+        {
+            # TODO: run this within a job to speed up the removal
+            $extensionName = $extension.Id.split("/") | select -Last 1
+
+             $jobs += Start-Job  -Name "Remove-CSX-$($vm.Name)-$($extensionName)" –Scriptblock {
```

Comment:
Unfortunately, i don't believe that this style is supported with powershell script blocks.